zerver/views/users.py

```diff
@@ -228,8 +234,25 @@ def avatar(
     except ValueError:
         is_email = True
 
+    if not maybe_user_profile.is_authenticated:
+        # Allow anonynous access to avatars only if spectators are
+        # enabled in the organization.
+        realm = get_valid_realm_from_request(request)
+        # TODO: Replace with realm.allow_web_public_streams_access()
+        # when the method is available.
+        if not realm.has_web_public_streams():
+            raise MissingAuthenticationError()
+
+        # We only allow the ID format for accessing a user's avatar
+        # for spectators. This is mainly for defense in depth, since
+        # email_address_visibility should mean spectators only
+        # interact with fake email addresses anyway.
+        if is_email:
+            raise MissingAuthenticationError()
```

Comment:
I added this `is_email` check -- under the theory that if we're not careful, this logic could be used to try to brute-force users' email addresses if somehow `email_address_visibility` isn't configured properly.

Reply:
Actually, we use `avatar/email/medium` url for `user_profile_modal.hbs` which is visible on clicking `View full profile`. Since we plan to not display the modal for spectators this change doesn't break anything. Although, we still get an 401 message in console by just clicking on user name in message due to this change. So, If there is no specific reason for using email to get `user_avatar` in `user_profile_modal.hbs`, should I change that to use `id` instead of `email`?

Reply:
I figured out the additional avatar using email fetch is due to [this](https://github.com/zulip/zulip/blob/61c082503668b9b20658530c978f2d9d1a7460a6/static/js/popovers.js#L259). We fetch the avatar image twice here, first when rendering the template and then in `load_medium_avatar` at end. I am not sure what the history is here?

Reply:
We should just change those calls to fetch by user ID rather than email, right?  

I think the background may have been that we want an image to appear immediately -- and we basically know the user's small avatar is in cache -- so it'll show a low-resolution immediately and then transparently swap to the higher resolution once it gets a response from the network.  We should add a comment clarifying this.

Reply:
I'll add said comment and try to do the cutover away from using /email/ URLs with avatars in the frontend codebase in general.